server/config.go

```diff
@@ -153,13 +154,14 @@ type Config struct {
 	MattermostWebhookFooter string
 
 	LogSettings struct {
-		EnableConsole bool
-		ConsoleJSON   bool
-		ConsoleLevel  string
-		EnableFile    bool
-		FileJSON      bool
-		FileLevel     string
-		FileLocation  string
+		EnableConsole   bool
+		ConsoleJSON     bool
+		ConsoleLevel    string
+		EnableFile      bool
+		FileJSON        bool
+		FileLevel       string
+		FileLocation    string
+		AdvancedLogging mlog.LogTargetCfg
```

Comment:
Happy to add examples, but I'm not clear on what logging is desired, and whether config based set up will be used or programmatically adding targets will be done.

Reply:
@khos2ow - do you have any comments on this? How does this play with the new k8s setup? We now have the ability to redirect logs to a custom destination. Would such a feature be helpful in production?

Would it be beneficial to directly write to a socket [endpoint](https://www.elastic.co/guide/en/logstash/current/plugins-inputs-tcp.html) in logstash?

Reply:
I was just looking at this. Although having this specific flexibility is really cool but in the scope of having the app deployed on live on a cluster will not be that beneficial. The way extracting logs of a pod on a kubernetes works is that we usually deploy an agent along side it which looks for any of `stdout`, `stderr` or log file inside a pod, and pipes them to its own stdout or stderr which gets shipped to backend storage.

So as long as matterbuild writes logs to either of the three above (stdout and stderr are actually preferred for the sake of simplicity and prevent us to deal with rotating log file), then I'd be able to ship it in our logging system.

> Would it be beneficial to directly write to a socket endpoint in logstash?

No, this is gonna introduce a hard dependency on a technology (which actually in this exact case we are not using!). This is the job of log exporter, fleutnd or promtail (we use this), which in turn awares of logging backend system, ELK, EFK or loki (we use this).

By making the pod not aware of the underlying logging stack we can easily swap the said logging stack with other one as we might see fit.

Reply:
This type of logging is specifically designed to avoid having to deploy an agent to capture logs via stdout/stderr.  The agent is built-in to the logger replacing a thing that needs to be deployed and monitored.   Where will the agent be sending the logs?

Reply:
Installing a log agent is literally the first thing we do for any cluster after the cluster is created, so that's not something that you might worry that it might put some operational overhead if you didn't embed it into the app itself.

> Where will the agent be sending the logs?

I don't know, and I kind of don't care (to be frank I do know and I do care!). I mean that is a runtime operational decision, today we've decided on one specific technology on one specific backend on one specific storage type but in the future that can change and it shouldn't require a change in the app that produces those log.

Reply:
I was asked by the cloud team to include the ability to send logs to syslog via TLS for their k8s instances. Removing the need for these log forwarders within the container was specifically cited as one of the motivations.  If you don't want that for mattermod, that's fine, but for the main k8s clusters I hope that is not the case.  

Also, no change is required in the app to decide on a different log sink.  The built-in forwarder gets configured the same way a third party forwarder does.

Should I close this PR and ticket as unnecessary?

Reply:
I can see the point you raised from the cloud team, and I also can see the same point for mattermod too. It is in fact open source so anyone can opt to use it, as well as ourselves. So having this mechanism, universally consistent across different part of the whole product is nice.

I'm +1 on having this merged but 0 on using it in actual mattermod deployment.

Reply:
Thanks for the discussion everyone. The original objective of this PR was to forward the log output to a byte buffer to be able to compare test outputs. So it is absolutely necessary for this PR to go in. The point about using in production was a by-product of adding this API. I will leave the production decision to the devops team, but for the code, this is needed.

@wiggin77 - if you can fix the conflicts, I think this is good to go. :) Thanks for your patience with this.

Reply:
Merge conflicts are fixed.